File: python/src/parakeet_onnx/contracts.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import json
from pathlib import Path
import re
from typing import Any, Mapping


RUN_CONTEXT_SCHEMA_VERSION = 2
GENERATED_CANDIDATE_SCHEMA_VERSION = 1

_ENVIRONMENTS = frozenset({"linux", "windows", "macos"})
_PROVIDERS = frozenset({"cpu", "cuda", "directml", "coreml"})
_EVALUATIONS = frozenset({"smoke", "parity", "coreml-parity", "full"})
_IMPLEMENTATIONS = frozenset({"python", "rust"})
_BACKENDS = frozenset({"onnxruntime"})
_DECODERS = frozenset({"ctc", "tdt", "whisper_autoregressive"})
_INPUT_KINDS = frozenset({"canonical_waveform", "features"})
_SHA256_RE = re.compile(r"^[0-9a-fA-F]{64}$")
_GIT_RE = re.compile(r"^[0-9a-fA-F]{7,64}$")
_CONFIG_VERSION_RE = re.compile(r"^config-\d{6}$")


class ContractError(RuntimeError):
    pass


def require_nonempty(name: str, value: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ContractError(f"{name} must be a non-empty string")
    return value


def require_sha256(name: str, value: str) -> str:
    require_nonempty(name, value)
    if _SHA256_RE.fullmatch(value) is None:
        raise ContractError(f"{name} must be a 64-character SHA-256")
    return value.lower()


def require_git_commit(value: str) -> str:
    require_nonempty("git.commit", value)
    if _GIT_RE.fullmatch(value) is None:
        raise ContractError("git.commit must be a 7-64 character hexadecimal commit identity")
    return value.lower()


def require_config_version(value: str) -> str:
    require_nonempty("revisions.config_version", value)
    if _CONFIG_VERSION_RE.fullmatch(value) is None:
        raise ContractError("revisions.config_version must match config-NNNNNN")
    return value


def require_one_of(name: str, value: str, allowed: frozenset[str]) -> str:
    require_nonempty(name, value)
    if value not in allowed:
        raise ContractError(
            f"{name} has unsupported value {value!r}; expected one of {sorted(allowed)}"
        )
    return value


def reject_nulls(value: Any, path: str = "$") -> None:
    if value is None:
        raise ContractError(f"contract must not contain null: {path}")
    if isinstance(value, Mapping):
        for key, item in value.items():
            reject_nulls(item, f"{path}.{key}")
    elif isinstance(value, (list, tuple)):
        for index, item in enumerate(value):
            reject_nulls(item, f"{path}[{index}]")


@dataclass(frozen=True, slots=True)
class GeneratedCatalog:
    id: str
    sha256: str

    def validate(self) -> None:
        require_nonempty("candidate.catalog.id", self.id)
        require_sha256("candidate.catalog.sha256", self.sha256)


@dataclass(frozen=True, slots=True)
class GeneratedArtifact:
    path: str
    sha256: str
    size_bytes: int

    def validate(self, role: str) -> None:
        require_nonempty(f"candidate.artifacts.{role}.path", self.path)
        require_sha256(f"candidate.artifacts.{role}.sha256", self.sha256)
        if not isinstance(self.size_bytes, int) or isinstance(self.size_bytes, bool) or self.size_bytes <= 0:
            raise ContractError(
                f"candidate.artifacts.{role}.size_bytes must be a positive integer"
            )


@dataclass(frozen=True, slots=True)
class GeneratedTokenizer:
    kind: str
    path: str

    def validate(self) -> None:
        require_nonempty("candidate.tokenizer.kind", self.kind)
        require_nonempty("candidate.tokenizer.path", self.path)


@dataclass(frozen=True, slots=True)
class GeneratedRuntimeContract:
    decoder: str
    input_kind: str
    io: Mapping[str, Any]
    decoder_config: Mapping[str, Any]

    def validate(self) -> None:
        require_one_of("candidate.runtime_contract.decoder", self.decoder, _DECODERS)
        require_one_of("candidate.runtime_contract.input_kind", self.input_kind, _INPUT_KINDS)
        reject_nulls(self.io, "$.runtime_contract.io")
        reject_nulls(self.decoder_config, "$.runtime_contract.decoder_config")


@dataclass(frozen=True, slots=True)
class GeneratedCandidateContract:
    schema_version: int
    candidate_root: str
    candidate_id: str
    profile_set: str
    variant: str
    profile: str
    decoder: str
    artifact_contract: str
    catalog: GeneratedCatalog
    bundle_sha256: str
    artifacts: Mapping[str, GeneratedArtifact]
    tokenizer: GeneratedTokenizer | None
    features: Mapping[str, bool]
    runtime_contract: GeneratedRuntimeContract

    def validate(self) -> None:
        if self.schema_version != GENERATED_CANDIDATE_SCHEMA_VERSION:
            raise ContractError(
                f"generated candidate schema_version must equal {GENERATED_CANDIDATE_SCHEMA_VERSION}"
            )
        for name, value in (
            ("candidate_root", self.candidate_root),
            ("candidate_id", self.candidate_id),
            ("profile_set", self.profile_set),
            ("variant", self.variant),
            ("profile", self.profile),
            ("artifact_contract", self.artifact_contract),
        ):
            require_nonempty(f"candidate.{name}", value)
        require_one_of("candidate.decoder", self.decoder, _DECODERS)
        require_sha256("candidate.bundle_sha256", self.bundle_sha256)
        self.catalog.validate()
        if self.decoder != self.runtime_contract.decoder:
            raise ContractError(
                "candidate.decoder must equal candidate.runtime_contract.decoder"
            )
        if not self.artifacts:
            raise ContractError("candidate.artifacts must not be empty")
        for role, artifact in self.artifacts.items():
            require_nonempty("candidate artifact role", role)
            artifact.validate(role)
        if self.tokenizer is not None:
            self.tokenizer.validate()
        for key, value in self.features.items():
            require_nonempty("candidate feature name", key)
            if type(value) is not bool:
                raise ContractError(f"candidate.features.{key} must be boolean")
        self.runtime_contract.validate()
        reject_nulls(self.to_dict())

    def to_dict(self) -> dict[str, Any]:
        value: dict[str, Any] = {
            "schema_version": self.schema_version,
            "candidate_root": self.candidate_root,
            "candidate_id": self.candidate_id,
            "profile_set": self.profile_set,
            "variant": self.variant,
            "profile": self.profile,
            "decoder": self.decoder,
            "artifact_contract": self.artifact_contract,
            "catalog": asdict(self.catalog),
            "bundle_sha256": self.bundle_sha256,
            "artifacts": {
                role: asdict(artifact)
                for role, artifact in sorted(self.artifacts.items())
            },
            "features": dict(sorted(self.features.items())),
            "runtime_contract": {
                "decoder": self.runtime_contract.decoder,
                "input_kind": self.runtime_contract.input_kind,
                "io": dict(self.runtime_contract.io),
                "decoder_config": dict(self.runtime_contract.decoder_config),
            },
        }
        if self.tokenizer is not None:
            value["tokenizer"] = asdict(self.tokenizer)
        return value


@dataclass(frozen=True, slots=True)
class CatalogReference:
    id: str
    sha256: str

    def validate(self) -> None:
        require_nonempty("revisions.runtime.catalog.id", self.id)
        require_sha256("revisions.runtime.catalog.sha256", self.sha256)


@dataclass(frozen=True, slots=True)
class RepoRevisionIdentity:
    repo_id: str
    revision: str

    def validate(self, name: str) -> None:
        require_nonempty(f"{name}.repo_id", self.repo_id)
        require_nonempty(f"{name}.revision", self.revision)


@dataclass(frozen=True, slots=True)
class RuntimeRevisionSnapshot:
    document_sha256: str
    catalog: CatalogReference
    profile_set: str

    def validate(self) -> None:
        require_sha256("revisions.runtime.document_sha256", self.document_sha256)
        self.catalog.validate()
        require_nonempty("revisions.runtime.profile_set", self.profile_set)


@dataclass(frozen=True, slots=True)
class ReferenceRevisionSnapshot:
    document_sha256: str
    development_artifact: RepoRevisionIdentity
    upstream: RepoRevisionIdentity
    tokenizer: RepoRevisionIdentity
    reference_id: str
    reference_revision: str
    canonical_framework: str

    def validate(self) -> None:
        require_sha256("revisions.reference.document_sha256", self.document_sha256)
        self.development_artifact.validate("revisions.reference.development_artifact")
        self.upstream.validate("revisions.reference.upstream")
        self.tokenizer.validate("revisions.reference.tokenizer")
        require_nonempty("revisions.reference.reference_id", self.reference_id)
        require_nonempty("revisions.reference.reference_revision", self.reference_revision)
        require_nonempty("revisions.reference.canonical_framework", self.canonical_framework)


@dataclass(frozen=True, slots=True)
class EvaluationSchemaRevisionSnapshot:
    document_sha256: str
    schema_id: str
    schema_revision: str

    def validate(self) -> None:
        require_sha256("revisions.evaluation_schema.document_sha256", self.document_sha256)
        require_nonempty("revisions.evaluation_schema.schema_id", self.schema_id)
        require_nonempty("revisions.evaluation_schema.schema_revision", self.schema_revision)


@dataclass(frozen=True, slots=True)
class DatasetRevisionEntry:
    id: str
    repo_id: str
    revision: str
    subset: str
    split: str
    sha256: str
    manifest: str

    def validate(self, index: int) -> None:
        prefix = f"revisions.datasets.entries[{index}]"
        require_nonempty(f"{prefix}.id", self.id)
        require_nonempty(f"{prefix}.repo_id", self.repo_id)
        require_nonempty(f"{prefix}.revision", self.revision)
        require_nonempty(f"{prefix}.subset", self.subset)
        require_nonempty(f"{prefix}.split", self.split)
        require_sha256(f"{prefix}.sha256", self.sha256)
        require_nonempty(f"{prefix}.manifest", self.manifest)


@dataclass(frozen=True, slots=True)
class DatasetsRevisionSnapshot:
    document_sha256: str
    entries: tuple[DatasetRevisionEntry, ...]

    def validate(self) -> None:
        require_sha256("revisions.datasets.document_sha256", self.document_sha256)
        ids: set[str] = set()
        for index, entry in enumerate(self.entries):
            entry.validate(index)
            if entry.id in ids:
                raise ContractError(f"duplicate dataset revision id: {entry.id}")
            ids.add(entry.id)


@dataclass(frozen=True, slots=True)
class RevisionSnapshot:
    config_version: str
    bundle_sha256: str
    runtime: RuntimeRevisionSnapshot
    reference: ReferenceRevisionSnapshot
    evaluation_schema: EvaluationSchemaRevisionSnapshot
    datasets: DatasetsRevisionSnapshot

    def validate(self) -> None:
        require_config_version(self.config_version)
        require_sha256("revisions.bundle_sha256", self.bundle_sha256)
        self.runtime.validate()
        self.reference.validate()
        self.evaluation_schema.validate()
        self.datasets.validate()
        reject_nulls(self.to_dict())

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class ArtifactIdentity:
    path: str
    sha256: str
    size_bytes: int
    candidate_id: str
    artifact_role: str

    def validate(self) -> None:
        require_nonempty("artifact.path", self.path)
        require_sha256("artifact.sha256", self.sha256)
        require_nonempty("artifact.candidate_id", self.candidate_id)
        require_nonempty("artifact.artifact_role", self.artifact_role)
        if self.size_bytes <= 0:
            raise ContractError("artifact.size_bytes must be greater than zero")


@dataclass(frozen=True, slots=True)
class GitIdentity:
    repository: str
    commit: str
    ref: str
    dirty: bool

    def validate(self) -> None:
        require_nonempty("git.repository", self.repository)
        require_git_commit(self.commit)
        require_nonempty("git.ref", self.ref)
        if type(self.dirty) is not bool:
            raise ContractError("git.dirty must be boolean")


@dataclass(frozen=True, slots=True)
class HostIdentity:
    os: str
    architecture: str
    hostname: str
    python_version: str
    implementation: str
    is_wsl: bool
    github_runner_os: str
    github_runner_arch: str
    github_run_id: str
    github_run_attempt: str

    def validate(self) -> None:
        for name, value in (
            ("host.os", self.os),
            ("host.architecture", self.architecture),
            ("host.hostname", self.hostname),
            ("host.python_version", self.python_version),
            ("host.implementation", self.implementation),
            ("host.github_runner_os", self.github_runner_os),
            ("host.github_runner_arch", self.github_runner_arch),
            ("host.github_run_id", self.github_run_id),
            ("host.github_run_attempt", self.github_run_attempt),
        ):
            require_nonempty(name, value)
        if type(self.is_wsl) is not bool:
            raise ContractError("host.is_wsl must be boolean")


@dataclass(frozen=True, slots=True)
class RuntimeIdentity:
    implementation: str
    backend: str
    backend_version: str
    provider_id: str
    provider_ort_name: str
    provider_available: bool

    def validate(self) -> None:
        require_one_of("runtime.implementation", self.implementation, _IMPLEMENTATIONS)
        require_one_of("runtime.backend", self.backend, _BACKENDS)
        require_nonempty("runtime.backend_version", self.backend_version)
        require_one_of("runtime.provider_id", self.provider_id, _PROVIDERS)
        require_nonempty("runtime.provider_ort_name", self.provider_ort_name)
        if type(self.provider_available) is not bool:
            raise ContractError("runtime.provider_available must be boolean")


@dataclass(frozen=True, slots=True)
class ConfigSnapshot:
    identity: str
    sources: Mapping[str, str]
    resolved: Mapping[str, Any]

    def validate(self) -> None:
        require_nonempty("config.identity", self.identity)
        for required in ("model", "provider", "environment", "evaluation"):
            value = self.sources.get(required)
            require_nonempty(f"config.sources.{required}", value if isinstance(value, str) else "")
        reject_nulls(self.resolved, "$.config.resolved")


@dataclass(frozen=True, slots=True)
class RunContext:
    schema_version: int
    run_id: str
    created_at: str
    config_identity: str
    model_id: str
    environment_id: str
    provider_id: str
    evaluation_id: str
    artifact: ArtifactIdentity
    git: GitIdentity
    host: HostIdentity
    runtime: RuntimeIdentity
    revisions: RevisionSnapshot
    config: ConfigSnapshot
    metadata: Mapping[str, Any]

    def validate(self) -> None:
        if self.schema_version != RUN_CONTEXT_SCHEMA_VERSION:
            raise ContractError(
                f"run-context schema_version must equal {RUN_CONTEXT_SCHEMA_VERSION}"
            )
        for name, value in (
            ("run_id", self.run_id),
            ("created_at", self.created_at),
            ("config_identity", self.config_identity),
            ("model_id", self.model_id),
        ):
            require_nonempty(name, value)
        require_one_of("environment_id", self.environment_id, _ENVIRONMENTS)
        require_one_of("provider_id", self.provider_id, _PROVIDERS)
        require_one_of("evaluation_id", self.evaluation_id, _EVALUATIONS)
        self.artifact.validate()
        self.git.validate()
        self.host.validate()
        self.runtime.validate()
        self.revisions.validate()
        self.config.validate()
        if self.runtime.provider_id != self.provider_id:
            raise ContractError("runtime.provider_id must equal provider_id")
        candidate = self.metadata.get("candidate")
        if not isinstance(candidate, Mapping):
            raise ContractError("metadata.candidate is required")
        if candidate.get("candidate_id") != self.artifact.candidate_id:
            raise ContractError(
                "artifact.candidate_id must equal metadata.candidate.candidate_id"
            )
        if candidate.get("profile_set") != self.revisions.runtime.profile_set:
            raise ContractError(
                "metadata.candidate.profile_set must equal revisions.runtime.profile_set"
            )
        catalog = candidate.get("catalog")
        if not isinstance(catalog, Mapping):
            raise ContractError("metadata.candidate.catalog must be an object")
        if (
            catalog.get("id") != self.revisions.runtime.catalog.id
            or catalog.get("sha256") != self.revisions.runtime.catalog.sha256
        ):
            raise ContractError(
                "metadata.candidate.catalog must equal revisions.runtime.catalog"
            )
        reject_nulls(self.metadata, "$.metadata")
        reject_nulls(self.to_dict())

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    def to_json(self, *, indent: int = 2) -> str:
        self.validate()
        return json.dumps(
            self.to_dict(),
            ensure_ascii=False,
            indent=indent,
            sort_keys=True,
        )

    def write_json(self, path: str | Path) -> None:
        destination = Path(path)
        destination.parent.mkdir(parents=True, exist_ok=True)
        destination.write_text(self.to_json() + "\n", encoding="utf-8")

File: rust/crates/asr-contracts/tests/contracts.rs
use std::fs;
use std::time::{SystemTime, UNIX_EPOCH};

use asr_contracts::{
    validate_benchmark, validate_run_context, validate_run_directory, validate_sample_result,
};
use serde_json::{Value, json};

const SHA: &str = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

fn candidate() -> Value {
    json!({
        "schema_version":1,
        "candidate_root":"/candidate",
        "candidate_id":"candidate-000001",
        "profile_set":"parakeet-tdt-ctc-v1",
        "variant":"ctc",
        "profile":"ctc-v1",
        "decoder":"ctc",
        "artifact_contract":"ctc-single-graph-v1",
        "catalog":{"id":"asr-runtime-catalog-v1","sha256":SHA},
        "bundle_sha256":SHA,
        "artifacts":{"primary":{"path":"model.onnx","sha256":SHA,"size_bytes":1}},
        "features":{},
        "runtime_contract":{
            "decoder":"ctc",
            "input_kind":"canonical_waveform",
            "io":{"primary":{"input":"audio_signal","length_input":"length","logits_output":"logits"}},
            "decoder_config":{"blank_id":0}
        }
    })
}

fn run_context() -> Value {
    json!({
        "schema_version":2,
        "run_id":"run-1",
        "created_at":"2026-08-16T00:00:00+00:00",
        "config_identity":"model:linux:cpu:smoke",
        "model_id":"model",
        "environment_id":"linux",
        "provider_id":"cpu",
        "evaluation_id":"smoke",
        "artifact":{
            "path":"candidate/model.onnx","sha256":SHA,"size_bytes":1,
            "candidate_id":"candidate-000001","artifact_role":"primary"
        },
        "git":{"repository":"owner/repo","commit":"deadbeef","ref":"refs/heads/main","dirty":false},
        "host":{
            "os":"Linux","architecture":"x86_64","hostname":"runner",
            "python_version":"3.12.0","implementation":"CPython","is_wsl":false,
            "github_runner_os":"Linux","github_runner_arch":"X64",
            "github_run_id":"1","github_run_attempt":"1"
        },
        "runtime":{
            "implementation":"python","backend":"onnxruntime","backend_version":"1.28.0",
            "provider_id":"cpu","provider_ort_name":"CPUExecutionProvider","provider_available":true
        },
        "revisions":{
            "config_version":"config-000001","bundle_sha256":SHA,
            "runtime":{"document_sha256":SHA,"catalog":{"id":"asr-runtime-catalog-v1","sha256":SHA},"profile_set":"parakeet-tdt-ctc-v1"},
            "reference":{
                "document_sha256":SHA,
                "development_artifact":{"repo_id":"dev/model","revision":"dev"},
                "upstream":{"repo_id":"up/model","revision":"up"},
                "tokenizer":{"repo_id":"up/model","revision":"tok"},
                "reference_id":"reference-v1","reference_revision":"ref","canonical_framework":"nemo"
            },
            "evaluation_schema":{"document_sha256":SHA,"schema_id":"eval-v1","schema_revision":"eval"},
            "datasets":{"document_sha256":SHA,"entries":[]},
            "provenance":{"document_sha256":SHA,"manifest_sha256":SHA,"status":"complete","automation_consumption":true,"target_id":"parakeet-tdt_ctc-0.6b-ja"}
        },
        "config":{
            "identity":"model:linux:cpu:smoke",
            "sources":{
                "model":"config/models/model.toml","provider":"config/providers/cpu.toml",
                "environment":"config/environments/linux.toml","evaluation":"config/evaluation/smoke.toml"
            },
            "resolved":{
                "model":{},"provider":{},"environment":{},"evaluation":{},
                "resolved":{"model_id":"model","provider_id":"cpu","environment_id":"linux","evaluation_id":"smoke"}
            }
        },
        "metadata":{"candidate":candidate(),"runtime_variant":"ctc","runtime_profile":"ctc-v1","provenance":{"manifest_sha256":SHA,"status":"complete","automation_consumption":true,"target_id":"parakeet-tdt_ctc-0.6b-ja"}}
    })
}

fn sample() -> Value {
    json!({
        "schema_version":1,"run_id":"run-1",
        "sample":{
            "id":"sample-1","dataset_id":"dataset","dataset_repo_id":"org/dataset",
            "dataset_revision":"revision","subset":null,"split":"test","index":0,
            "audio_sha256":SHA,"audio_duration_sec":1.0,"sample_rate_hz":16000,"reference_text":"参照"
        },
        "execution":{"runtime":"rust","backend":"onnxruntime","provider_id":"cpu","decoder":"ctc","batch_size":1},
        "output":{"text":"認識","normalized_text":"認識","tokens":[1],"token_count":1},
        "quality":{"cer":0.1,"wer":0.2},
        "timing":{
            "load_ms":null,"session_creation_ms":null,"audio_decode_ms":1.0,"resample_ms":0.0,
            "frontend_ms":null,"encoder_ms":null,"decoder_ms":0.1,"postprocess_ms":0.1,
            "inference_ms":10.0,"total_ms":11.2,"rtf":0.0112
        },
        "memory":{"peak_ram_mb":64.0,"peak_device_memory_mb":null},
        "parity":{
            "reference_run_id":null,"text_match":null,"token_match":null,
            "numeric":{
                "frontend":{"compared":false,"passed":null,"max_abs_error":null,"mean_abs_error":null,"relative_l2":null},
                "encoder":{"compared":false,"passed":null,"max_abs_error":null,"mean_abs_error":null,"relative_l2":null},
                "logits":{"compared":false,"passed":null,"max_abs_error":null,"mean_abs_error":null,"relative_l2":null}
            }
        },
        "provider":{
            "requested":"cpu","registered":true,"used":true,"fallback_detected":false,
            "fallback_only":false,"assigned_nodes":null,"fallback_nodes":0
        },
        "status":"success","errors":[]
    })
}

fn benchmark() -> Value {
    json!({
        "schema_version":1,"run_id":"run-1",
        "candidate":{"candidate_id":"candidate-000001","model_id":"model","artifact_sha256":SHA,"artifact_size_bytes":1,"decoder":"ctc"},
        "evaluation":{
            "suite":"smoke","manifest":"evaluation/manifest.json","expected_sample_count":1,
            "reference_revision_sha256":SHA,"evaluation_schema_sha256":SHA,"datasets_lock_sha256":SHA,"revision_bundle_sha256":SHA
        },
        "runtime":{
            "implementation":"rust","backend":"onnxruntime","backend_version":"1.28.0",
            "environment_id":"linux","provider_id":"cpu","provider_ort_name":"CPUExecutionProvider",
            "os":"Linux","architecture":"x86_64"
        },
        "samples":{"expected":1,"attempted":1,"successful":1,"failed":0,"skipped":0,"total_audio_duration_sec":1.0},
        "quality":{"cer":0.1,"wer":0.2},
        "performance":{
            "load_ms":null,"session_creation_ms":1.0,"total_processing_ms":11.2,"rtf":0.0112,
            "per_sample":{"mean_ms":11.2,"median_ms":11.2,"p50_ms":11.2,"p95_ms":11.2,"p99_ms":11.2,"min_ms":11.2,"max_ms":11.2},
            "components":{"audio_decode_ms":1.0,"resample_ms":0.0,"frontend_ms":null,"encoder_ms":null,"decoder_ms":0.1,"postprocess_ms":0.1,"inference_ms":10.0}
        },
        "memory":{"peak_ram_mb":64.0,"peak_device_memory_mb":null},
        "parity":{
            "reference_run_id":null,"text_matches":0,"text_mismatches":0,"token_matches":0,"token_mismatches":0,
            "text_match_rate":null,"token_match_rate":null,
            "numeric":{
                "frontend":{"compared_samples":0,"failed_samples":0,"max_abs_error":null,"max_mean_abs_error":null,"max_relative_l2":null},
                "encoder":{"compared_samples":0,"failed_samples":0,"max_abs_error":null,"max_mean_abs_error":null,"max_relative_l2":null},
                "logits":{"compared_samples":0,"failed_samples":0,"max_abs_error":null,"max_mean_abs_error":null,"max_relative_l2":null}
            }
        },
        "provider":{"requested":"cpu","registered":true,"execution_proven":true,"fallback_detected":false,"fallback_only":false,"assigned_nodes":null,"fallback_nodes":0},
        "acceptance":{"passed":true,"quality_passed":true,"parity_passed":null,"provider_passed":true,"performance_passed":true,"failed_checks":[],"warnings":[]},
        "errors":{"total":0,"fatal":0,"by_code":{}}
    })
}

#[test]
fn validates_current_canonical_contract_shapes() {
    validate_run_context(&run_context()).unwrap();
    validate_sample_result(&sample()).unwrap();
    validate_benchmark(&benchmark()).unwrap();
}

#[test]
fn run_context_rejects_candidate_profile_set_mismatch() {
    let mut value = run_context();
    value["metadata"]["candidate"]["profile_set"] = Value::String("other".into());
    assert!(validate_run_context(&value).is_err());
}

#[test]
fn run_directory_validation_cross_checks_run_and_samples() {
    let unique = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    let dir = std::env::temp_dir().join(format!("jpapt-contracts-{}-{unique}", std::process::id()));
    fs::create_dir_all(&dir).unwrap();
    fs::write(
        dir.join("run-context.json"),
        serde_json::to_vec_pretty(&run_context()).unwrap(),
    )
    .unwrap();
    fs::write(
        dir.join("metrics.json"),
        serde_json::to_vec_pretty(&benchmark()).unwrap(),
    )
    .unwrap();
    fs::write(
        dir.join("samples.jsonl"),
        format!("{}\n", serde_json::to_string(&sample()).unwrap()),
    )
    .unwrap();

    let summary = validate_run_directory(&dir).unwrap();
    assert_eq!(summary.run_id, "run-1");
    assert_eq!(summary.sample_count, 1);

    fs::remove_dir_all(dir).unwrap();
}
